Feature/UTC-248-Pattern-Lab-Buttons

brittrich opened the pull request:
This adds the buttons to Pattern Lab. I'll add a few notes to the code with questions/clarifications:

![pl-demo](https://user-images.githubusercontent.com/50490141/131068068-b5444cdb-6639-4eec-88c1-2a40d0eac6fb.gif)

One question I have about PL in general is whether it would be useful to have all the template options visible in the twig instead of what we are naming in the .yml? So when building other components it is easier to see what options are available to be modified. I tried a bit to get this to work, but couldn't solve it.

<img width="1501" alt="Screen Shot 2021-08-26 at 11 36 10 PM" src="https://user-images.githubusercontent.com/50490141/131068219-492398fd-4922-4b8c-84d9-ccda166143b7.png">

Also, will you let me know if there are any additional files that need to be added in a commit like this?

I plan on doing some additional testing before finalizing the PR, just wanted to go ahead and let you look at it for feedback!

Commit:
add option for button vs. button link

Commit:
add button colors

Commit:
start to button components

Commit:
add button component base

Commit:
tweak sizing options

Commit:
button sizing files

Commit:
add button color options

Commit:
add button sizes

Commit:
add button styles

Commit:
more updates to styles

Commit:
minor edits for style

brittrich commented on apps/pl-default/pattern-lab/_patterns/00-protons-demo/_demo-block.twig:
```diff
@@ -18,7 +18,7 @@
         {% if demo.title %}
-          <h1 class="text-6xl mb-3">
+          <h1 class="text-4xl mb-3">
             {{ demo.title|title }}
```
We don't have to do this, I just felt like the h1 size was overwhelming.

UTCbridget replied:
Totally agree. I thought I had changed that, actually.

brittrich replied:
Maybe you did and I didn't pull in all the changes yet!

brittrich commented on dist/app-drupal/assets/atomic/_patterns/01-atoms/button/_button.twig:
```diff
@@ -14,25 +14,39 @@
+  %}
+
+{% endif %}
 
```
There may be a cleaner way to get the differentiation between the button and link, open to ideas!

brittrich commented on dist/app-drupal/assets/atomic/_patterns/01-atoms/button/_button.twig:
```diff
@@ -41,6 +55,4 @@
   {{ button.text }}
-  {% block right %}
-  {% endblock right %}
 </{{button_element}}>
```
Is this "block right" okay to remove? I couldn't figure out the purpose of it.

brittrich commented on source/default/_patterns/00-protons/legacy/css/components/buttons/_buttons.css:
```diff
@@ -172,12 +174,13 @@
+*/
 
 .btn--pill {
   border-radius: 24px;
```
We could possibly delete this .css file if the other component extraction works, left it w/ styles commented out for an easy revert if needed.

brittrich commented on source/default/_patterns/00-protons/tailwind.tokens.css:
```diff
@@ -14,10 +14,81 @@ Example: Google Remote Fonts
+@layer components {
+  .btn {
+    @apply text-center align-middle select-none mb-4 font-semibold py-1.5 px-3 text-base;
+  }
```
So, the buttons seemed like a unique use case and worth doing extracted components in tailwind. The button .css is shared with twig button use (e.g. cards) and also with buttons added in the text editors.

bmartinez287 replied:
makes sense although i wonder if we should prefix or use a different word since that's the bootstrap standard. It could be confusing for someone looking at it.

brittrich replied:
I guess I was trying to match what we've already done in _button.css as far as naming and styles...so that it doesn't have a site-wide impact (e.g. .btn in use in text editor). What do you think, am I thinking about it correctly?

bmartinez287 replied:
that's an interesting take. I had not considered its backward compatibility. It makes sense, I will read and think a bit more to see if there's more info about it.

brittrich replied:
Thanks for checking, @bmartinez287, just lmk if any changes are needed on this. We could always leave the bootstrap button classes in the css file, just wasn't sure about maintaining in two places. We can also take out these and just keep the button styles in .css if that's more logical with how we have things set up.

brittrich commented on source/default/_patterns/00-protons/tailwind.tokens.css:
```diff
@@ -14,10 +14,81 @@ Example: Google Remote Fonts
+
+  .btn--extra-large {
+    @apply p-6 text-xl font-bold font-utcheadings;
+   }
```
I couldn't figure out how to get Exo in here, but the large button block is meant to use it. It's assigned as utcheadings in the font-family token file, so maybe it has to do with an import?

bmartinez287 replied:
so font-utcheadings; did not work?

brittrich replied:
No, it didn't apply Exo when I used font-utcheadings (which I saw defined in the font-family file). I tried to figure out why but couldn't crack it.

bmartinez287 replied:
https://tailwindcss.com/docs/extracting-components

bmartinez287 replied:
It may be the order in which they were place maybe? 
`.btn--extra-large {
  @apply font-bold font-utcheadings p-6 text-xl ;
 }` 
I got it working with the text above as seen in the image below.
![Screen Shot 2021-08-30 at 2 24 29 PM](https://user-images.githubusercontent.com/39039024/131386701-86767c19-c646-42b7-8b74-c8487bedcfaa.png)

bmartinez287 replied:
@brittrich

UTCbridget replied:
I had the same font application problem in working with the PL demo for the cards. It wasn't picking up the load so I manually placed it in the twig file as an additional import. I will look at Bernardo's solution.

brittrich replied:
Hmm. So, I tried reordering and it still didn't work. I added the Exo import to the tailwind tokens css file and that seems to have kicked it in everywhere...here and the headings as well. Think that's the right spot for it? I saw Open Sans import but no Exo so figured it was worth a try to add it.

UTCbridget commented:
This looks great, Brittany! I'm going to put on my radar to incorporate these into the PL cards as includes.

Commit:
Merge remote-tracking branch 'upstream/develop' into feature/PL-buttons

Commit:
update base button description

Commit:
comment out button size css

Commit:
update button color description

bmartinez287 commented:
In regards to the first question at the beginning of this PR. How would that look like? Do you have a quick example of the difference?

Commit:
bolder button font

Commit:
fix inline on button style

Commit:
space change

brittrich commented:
> In regards to the first question at the beginning of this PR. How would that look like? Do you have a quick example of the difference?

Something like this? https://boltdesignsystem.com/pattern-lab/?p=elements-button 

When I looked at their git code it's definitely a different set up in general, but they do specify a block that designates which part of the template they want to use as the example.

bmartinez287 commented:
interesting, that's an alternate naming convention to what we are using. It makes sense to me. Especially, since those semantic names have more meaning once we use them.

Commit:
more rounded buttons

Commit:
Merge remote-tracking branch 'upstream/develop' into feature/PL-buttons

Commit:
add exo import

brittrich commented on source/default/_patterns/00-protons/tailwind.tokens.css:
```diff
@@ -14,9 +14,81 @@ Example: Google Remote Fonts
 
 @import url(https://fonts.googleapis.com/css?family=OpenSans|Oswald|SourceCodePro&display=swap);
+@import url(https://fonts.googleapis.com/css2?family=Exo:wght@700&display=swap);
+
```
Add Exo import here @bmartinez287 and @UTCbridget

UTCbridget replied:
Works for me! That's what I had to do for the cards.

brittrich commented:
@UTCbridget and @bmartinez287 I think these are ready for final review. I'm also fine if we want to wait to merge this PR until after the cards and/or hero PRs are merged.

UTCbridget commented:
Really nice job, Brittany! A lot of hard work!

bmartinez287 commented:
This looks great, yeah we can definitely talk about it tomorrow.

Commit:
adjust button hover contrast, padding

Commit:
add small and large bootstrap classes

Commit:
fix bold

brittrich commented:
I made a few minor edits to the button styles to improve hover indication:

1. Updated the hover color on the navy and yellow buttons where the contrast was lowest.
2. Added light shadow on hover for all button types.
3. Added slightly slower hover animation for all buttons.
4. Added more right and left padding to all button sizes.

**Before:**
![buttons-old](https://user-images.githubusercontent.com/50490141/133020377-9fda602a-b0cd-43ec-ab09-29d3b3191b1e.gif)

**After:**
![buttons-new](https://user-images.githubusercontent.com/50490141/133020402-c9303c5d-bda2-45c7-b496-5a263965e05c.gif)
